File: apps/web/src/types/delivery_renewal.ts
// Mirrors apps/api/app/schemas/delivery_renewal.py.

export type ExpandStage =
  | "value_proof"
  | "expand_ask"
  | "new_scope"
  | "close";

export type RedFlagType =
  | "missed_checkpoint"
  | "spoc_unresponsive"
  | "no_value_logged"
  | "escalation";

export type Outcome = "renewed" | "at_risk" | "not_renewed" | "undecided";

export type ReadinessAnswerValue = "yes" | "no" | "unknown";

export interface ExpandItem {
  id?: string | null;
  name: string;
  stage: ExpandStage;
  amount_musd?: number | null;
  note?: string | null;
}

export interface RedFlag {
  id?: string | null;
  type: RedFlagType;
  note?: string | null;
  raised_at?: string | null;
  raised_by?: string | null;
  resolved_at?: string | null;
  resolved_by?: string | null;
}

export interface ReadinessAnswer {
  answer: ReadinessAnswerValue;
  proof_note?: string | null;
}

export interface Readiness {
  delivered_metric: ReadinessAnswer;
  proof_data: ReadinessAnswer;
  client_acknowledged: ReadinessAnswer;
}

export interface Track1Derived {
  next_type: string | null;
  next_scheduled: string | null;
  next_days_until: number | null;
  overdue_count: number;
  signed_off_count: number;
  total: number;
}

export interface DeliveryRenewal {
  account_id: string;
  expand_value_proof: ExpandItem[];
  expand_expand_ask: ExpandItem[];
  expand_new_scope: ExpandItem[];
  expand_close: ExpandItem[];
  red_flags: RedFlag[];
  readiness: Readiness;
  track1: Track1Derived;
  expand_paused: boolean;
  readiness_score: number;
  outcome: Outcome | null;
  outcome_set_at: string | null;
  outcome_set_by: string | null;
  is_editable: boolean;
}

export interface DeliveryRenewalUpdate {
  expand_value_proof?: ExpandItem[] | null;
  expand_expand_ask?: ExpandItem[] | null;
  expand_new_scope?: ExpandItem[] | null;
  expand_close?: ExpandItem[] | null;
  readiness?: Readiness | null;
}

export const STAGES: ExpandStage[] = [
  "value_proof",
  "expand_ask",
  "new_scope",
  "close",
];

export const STAGE_LABELS: Record<ExpandStage, string> = {
  value_proof: "Value Proof",
  expand_ask: "Expand Ask",
  new_scope: "New Scope",
  close: "Close",
};

export const STAGE_COL_KEYS: Record<
  ExpandStage,
  | "expand_value_proof"
  | "expand_expand_ask"
  | "expand_new_scope"
  | "expand_close"
> = {
  value_proof: "expand_value_proof",
  expand_ask: "expand_expand_ask",
  new_scope: "expand_new_scope",
  close: "expand_close",
};

export const FLAG_TYPES: RedFlagType[] = [
  "missed_checkpoint",
  "spoc_unresponsive",
  "no_value_logged",
  "escalation",
];

export const FLAG_LABELS: Record<RedFlagType, string> = {
  missed_checkpoint: "Missed checkpoint",
  spoc_unresponsive: "SPOC unresponsive",
  no_value_logged: "No value logged",
  escalation: "Escalation raised",
};

export const OUTCOME_LABELS: Record<Outcome, string> = {
  renewed: "Renewed",
  at_risk: "At risk",
  not_renewed: "Not renewed",
  undecided: "Undecided",
};

/** Brand-palette colours per outcome (Beroe brand book Sept 2025).
 *  Hex values so callers can use inline styles. Mapping:
 *    renewed     → Risk Green #6EC457
 *    at_risk     → Risk Amber #F0BC41
 *    not_renewed → Risk Red   #CF4548
 *    undecided   → brand grey #94a3b8
 */
export const OUTCOME_TONES: Record<
  Outcome,
  { bg: string; border: string; text: string }
> = {
  renewed:     { bg: "#6EC45715", border: "#6EC45740", text: "#1d6b35" },
  at_risk:     { bg: "#F0BC4115", border: "#F0BC4140", text: "#854F0B" },
  not_renewed: { bg: "#CF454810", border: "#CF454830", text: "#7F1D1D" },
  undecided:   { bg: "#94a3b815", border: "#94a3b830", text: "#475569" },
};

export const READINESS_QUESTIONS: Array<{
  key: keyof Readiness;
  label: string;
  hint: string;
}> = [
  {
    key: "delivered_metric",
    label: "Did we deliver the metric?",
    hint: "Compare current vs target on the agreed success metric.",
  },
  {
    key: "proof_data",
    label: "Can we prove it with data?",
    hint: "Dashboard, audit log, signed-off checkpoint snapshot, etc.",
  },
  {
    key: "client_acknowledged",
    label: "Does the client acknowledge it?",
    hint: "Written confirmation, MBR sign-off, or email on file.",
  },
];

File: apps/api/app/schemas/delivery_renewal.py
"""M23 — Delivery & Renewal schemas.

Dual-track post-delivery view:
  * Track 1 (Renewal) — derived from M21 Checkpoints (next due, overdue
    count); the route hydrates `track1` from the checkpoints table.
  * Track 2 (Expand) — 4-stage Kanban (value_proof / expand_ask /
    new_scope / close). Stored as a list per stage.

Red flags pause Track 2. Renewal Readiness is a 3-question yes/no
assessment that gates the final renewal call. Outcome (renewed /
at_risk / not_renewed) is a stamped final decision — admin-only
re-open mirrors M13/M19/M22 unlock asymmetry.

`extra="allow"` everywhere keeps the shape elastic — same pattern as
M15 cs_goals + M19 success_contract + M22 vdd.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

ExpandStage = Literal["value_proof", "expand_ask", "new_scope", "close"]
RedFlagType = Literal[
    "missed_checkpoint",
    "spoc_unresponsive",
    "no_value_logged",
    "escalation",
]
Outcome = Literal["renewed", "at_risk", "not_renewed", "undecided"]


class ExpandItem(BaseModel):
    """One opportunity in the expand pipeline."""

    model_config = ConfigDict(extra="allow")

    id: str | None = None
    name: str
    stage: ExpandStage
    amount_musd: float | None = None
    note: str | None = None


class RedFlag(BaseModel):
    """One red flag. `resolved_at` null → still open → expand auto-pauses."""

    model_config = ConfigDict(extra="allow")

    id: str | None = None
    type: RedFlagType
    note: str | None = None
    raised_at: datetime | None = None
    raised_by: UUID | None = None
    resolved_at: datetime | None = None
    resolved_by: UUID | None = None


class ReadinessAnswer(BaseModel):
    """One of three renewal-readiness questions."""

    model_config = ConfigDict(extra="allow")

    answer: Literal["yes", "no", "unknown"] = "unknown"
    proof_note: str | None = None


class Readiness(BaseModel):
    """The 3 questions stored together so the score derives cleanly."""

    model_config = ConfigDict(extra="allow")

    delivered_metric: ReadinessAnswer = Field(default_factory=ReadinessAnswer)
    proof_data: ReadinessAnswer = Field(default_factory=ReadinessAnswer)
    client_acknowledged: ReadinessAnswer = Field(default_factory=ReadinessAnswer)


class Track1Derived(BaseModel):
    """Derived from M21 checkpoints. Read-only — recomputed every GET."""

    model_config = ConfigDict(extra="allow")

    next_type: str | None = None
    next_scheduled: str | None = None
    next_days_until: int | None = None
    overdue_count: int = 0
    signed_off_count: int = 0
    total: int = 0


class DeliveryRenewalOut(BaseModel):
    model_config = ConfigDict(extra="allow")

    account_id: UUID

    # Track 2 (Expand) — list per stage. Easier than an items[] + stage
    # because the UI is a 4-column Kanban; per-stage lists serialise 1:1
    # to columns.
    expand_value_proof: list[ExpandItem] = Field(default_factory=list)
    expand_expand_ask: list[ExpandItem] = Field(default_factory=list)
    expand_new_scope: list[ExpandItem] = Field(default_factory=list)
    expand_close: list[ExpandItem] = Field(default_factory=list)

    red_flags: list[RedFlag] = Field(default_factory=list)
    readiness: Readiness = Field(default_factory=Readiness)

    # Derived — recomputed each GET.
    track1: Track1Derived = Field(default_factory=Track1Derived)
    expand_paused: bool = False
    readiness_score: int = 0  # 0–3 count of `yes` answers

    outcome: Outcome | None = None
    outcome_set_at: datetime | None = None
    outcome_set_by: UUID | None = None

    is_editable: bool = False


class DeliveryRenewalUpdate(BaseModel):
    """Partial whole-document update. Refuses when an outcome is final
    unless the caller is admin (route enforces)."""

    model_config = ConfigDict(extra="allow")

    expand_value_proof: list[ExpandItem] | None = None
    expand_expand_ask: list[ExpandItem] | None = None
    expand_new_scope: list[ExpandItem] | None = None
    expand_close: list[ExpandItem] | None = None
    readiness: Readiness | None = None


class RedFlagCreate(BaseModel):
    model_config = ConfigDict(extra="allow")

    type: RedFlagType
    note: str | None = Field(None, max_length=600)


class OutcomeSet(BaseModel):
    """Set the final outcome. Once set, regular writers can't PATCH the
    rest of the document — only admin re-open releases."""

    model_config = ConfigDict(extra="allow")

    outcome: Outcome
